go/test/endtoend/vreplication/multi_tenant_test.go

```diff
@@ -0,0 +1,454 @@
+	vschemapb "vitess.io/vitess/go/vt/proto/vschema"
+)
+
+type tenantMigrationStatus int
+
+const (
+	tenantMigrationStatusNotMigrated tenantMigrationStatus = iota
+	tenantMigrationStatusMigrating
+	tenantMigrationStatusMigrated
+
+	sourceKeyspaceTemplate      = "s%d"
+	sourceAliasKeyspaceTemplate = "a%d"
+	targetKeyspaceName          = "mt"
+
+	numTenants                 = 10
+	numInitialRowsPerTenant    = 10
+	numAdditionalRowsPerTenant = 10
+	baseInitialTabletId        = 1000
+	tabletIdStep               = 100
+	maxRandomDelaySeconds      = 5
+	waitTimeout                = 10 * time.Minute
+)
+
+var (
+	// channels to coordinate the migration workflow
+	chNotSetup, chNotCreated, chInProgress, chSwitched, chCompleted chan int64
+	// counters to keep track of the number of tenants in each state
+	numSetup, numInProgress, numSwitched, numCompleted atomic.Int64
+)
+
+// multiTenantMigration manages the migration of multiple tenants to a single target keyspace.
+// A singleton object of this type is created for the test case.
+type multiTenantMigration struct {
+	t                     *testing.T
+	mu                    sync.Mutex
+	tenantMigrationStatus map[int64]tenantMigrationStatus // current migration status for each tenant
+	activeMoveTables      map[int64]*VtctldMoveTables     // the internal MoveTables object for each tenant
+
+	targetKeyspace     string
+	tables             string
+	tenantIdColumnName string // the name of the column in each table that holds the tenant ID
+
+	lastIDs map[int64]int64 // the last primary key inserted for each tenant
+}
+
+const (
+	mtSchema  = "create table t1(id int, tenant_id int, primary key(id, tenant_id)) Engine=InnoDB"
+	mtVSchema = `
+{
+  "multi_tenant_spec": {
+      "tenant_id_column_name": "tenant_id",
+      "tenant_id_column_type": "int64"
+  },
+  "tables": {
+    "t1": {}
+  }
+}
+`
+)
+
+// TestMultiTenantSimple tests a single tenant migration. The aim here is to test all the steps of the migration process
+// including keyspace routing rules, addition of tenant filters to the forward and reverse vreplication streams, and
+// verifying that the data is migrated correctly.
+func TestMultiTenantSimple(t *testing.T) {
+	setSidecarDBName("_vt")
+	// Don't create RDONLY tablets to reduce number of tablets created to reduce resource requirements for the test.
+	origDefaultRdonly := defaultRdonly
+	defer func() {
+		defaultRdonly = origDefaultRdonly
+	}()
+	defaultRdonly = 0
+	vc = setupMinimalCluster(t)
+	defer vc.TearDown()
+
+	targetKeyspace := "mt"
+	_, err := vc.AddKeyspace(t, []*Cell{vc.Cells["zone1"]}, targetKeyspace, "0", mtVSchema, mtSchema, 1, 0, 200, nil)
+	require.NoError(t, err)
+
+	tenantId := int64(1)
+	sourceKeyspace := getSourceKeyspace(tenantId)
+	sourceAliasKeyspace := getSourceAliasKeyspace(tenantId)
+	_, err = vc.AddKeyspace(t, []*Cell{vc.Cells["zone1"]}, sourceKeyspace, "0", mtVSchema, mtSchema, 1, 0, getInitialTabletIdForTenant(tenantId), nil)
```

Comment:
It seems that we require the same `multi_tenant_spec` in both source keyspace and target keyspace. Is this temporary during MoveTables, or expected to stay around forever?

Reply:
We only need the `multi_tenant_spec`  vschema for the target keyspace. Updated the e2e tests to remove it for the source keyspaces and confirmed locally as well.